libpysal/weights/network.py

```diff
@@ -0,0 +1,270 @@
+                    a dataframe with a geometry column that can be used to
+                    construct a W object
+        geom_col :   string
+                    column name of the geometry stored in df
+        network:    pandana.Network
+                    a pandana Network object (optionally created by `multimodal_from_bbox`). If none, the geodataframe's total_bounds
+                    attribute will be used to download an openstreetmap pedestrian network covering the study area.
+        ids     :   string or iterable
+                    if string, the column name of the indices from the dataframe
+                    if iterable, a list of ids to use for the W
+                    if None, pandana node_ids are used
+        max_dist:   int
+                    maximum distance along the road network to be considered neighbors
+
+        See Also
+        --------
+        :class:`libpysal.weights.weights.W`
+        """
+        if not network:
+            try:
+                from pandana.loaders import osm
+            except ImportError:
+                raise ImportError('You must have pandana installed to generate a Network')
+            assert 'epsg:4326' in df.crs.to_string(), "You must pass in an explicit pandana.Network object or provide a geodataframe in geographic (epsg:4326) coordinates"
+            network = osm.pdna_network_from_bbox(bbox=tuple(df.total_bounds))
+
+        if not geom_col == 'geometry':
+            df = df.set_geometry(geom_col)
+        adj = compute_travel_cost_adjlist(df, df, network, index_orig=ids, index_dest=ids)
+        if max_dist:
+            adj = adj[adj['cost'] <= max_dist]
+        return W.from_adjlist(adj, focal_col='origin', neighbor_col='destination', weight_col='cost')
+
+
+def compute_travel_cost_adjlist(
+    origins, destinations, network, index_orig=None, index_dest=None
+):
+    """Generate travel cost adjacency list.
+
+    Parameters
+    ----------
+    origins : geopandas.GeoDataFrame
+        a geodataframe containing the locations of origin features
+    destinations : geopandas.GeoDataFrame
+        a geodataframe containing the locations of destination features
+    network : pandana.Network
+        pandana Network instance for calculating the shortest path between origins and destinations
+    index_orig : str, optional
+        Unique index on the origins dataframe.
+    index_dest : str, optional
+        Unique index on the destinations dataframe.
+    Returns
+    -------
+    pandas.DataFrame
+        pandas DataFrame containing the shortest-cost distance from each origin feature to each destination feature
+    """
+    origins = origins.copy()
+    destinations = destinations.copy()
+
+    origins["osm_ids"] = network.get_node_ids(
+        origins.centroid.x, origins.centroid.y
+    ).astype(int)
+    destinations["osm_ids"] = network.get_node_ids(
+        destinations.centroid.x, destinations.centroid.y
+    ).astype(int)
```

Comment:
We could check if `origins == destinations` and copy `osm_ids` in that case. I guess that this can be costly operation, so it is worth skipping if we can.

Reply:
> do we know what is the performance of pandana's shortest paths vs networkx? I guess it is much faster since it is vectorised C code, but I haven't found it. 

i think i can find it somewhere. Its not only that pandana is vectorized c, but it also uses contraction hierarchies instead of e.g. dijkstra

Reply:
>We could check if origins == destinations and copy osm_ids in that case. I guess that this can be costly operation, so it is worth skipping if we can. 

thats a really good point. I originally wrote this for use [with access](https://github.com/knaaptime/healthacc), where origins and destinations aren't always the same, but in think in the context of a `W`, they *are*, so we can skip the second call to `get_node_ids`
```suggestion
    origins["osm_ids"] = network.get_node_ids(
        origins.centroid.x, origins.centroid.y
    ).astype(int)
    destinations["osm_ids"] = network.get_node_ids(
        destinations.centroid.x, destinations.centroid.y
    ).astype(int)
```